Python_DataEngineering/Extract_Store_Fetch.py

```diff
@@ -0,0 +1,158 @@
+        for properties in bikedata["additionalProperties"]: #here i am traversing json object present inside additionalProperties to check "Key" as "NbBikes" if it is available i am trying to extract the value associated with that key
+            if properties["key"] == "NbBikes":
+                IST=convertTOIST(properties["modified"])  #It passes the UTC Time as a String to convertTOIST function and gets IST Date and time as a String
+                bikepointname=bikedata["commonName"]
+                datetime=IST
+                noofbikes=properties["value"]
+                BikePointDetails[bikepointname]=[noofbikes,datetime]
+                #print("Bike Point Name: "+bikepointname)
+                #print("No oF Bikes Available: "+noofbikes)
+                #print("Date: "+IST)
+                #print(" ")
+    return BikePointDetails
+
+#bikesPerTimeSnapShot function traverses the API List and returns  the Total No. of Bikes available across London including all the BikePoints at a particular Point of Time
+def bikesPerTimeSnapShot(json_data):
+    timing_snap_shot={}
+    for bikedata in json_data:
+        for properties in bikedata["additionalProperties"]:
+            if properties["key"] == "NbBikes":
+                IST=convertTOIST(properties["modified"]) #this function converts UTC to IST and returns the result as a String
+                if IST not in timing_snap_shot:
+                    timing_snap_shot[IST]=int(properties["value"])
+                else:
+                    timing_snap_shot[IST]+=int(properties["value"])
+    return timing_snap_shot
+
+
+#This function inserts all the required values from API into twoo table
+#using sql1 table we will insert bikepointname, no of available bikes, datetime
+#using sql2 table we will insert total no of bikes present at a particular timesnapshot
+def databaseInsert(cursor,db,time_snap_shot_bike_data,bike_point_details):
+    sql1="insert into Bike_Detail(bikepointname, noofbike, datetime) VALUES (%s, %s, %s)"
+
+    #inserting into Total_Bike_Table using sql2
+    sql2="insert into time_snap_shot_detail(datetime, noofbike) VALUES (%s, %s)"
+    for date_time in time_snap_shot_bike_data:
+        datetime=date_time
+        no_of_bikes=str(time_snap_shot_bike_data[date_time])
+        value=(datetime,no_of_bikes)
+        cursor.execute(sql2,value)
+        db.commit()
+        
+    
+    
+    #inserting into Bike Table using sql1
+    for bikepoint in bike_point_details:
+        bikepointname=bikepoint
+        noofbikes=bike_point_details[bikepoint][0]
+        datetime=bike_point_details[bikepoint][1]
+        value=(bikepointname,noofbikes,datetime)
+        cursor.execute(sql1,value)
+        db.commit()
+        
+    
+     
+#this function will fetch no of bikes available as per user's selection of bikepointname and currenttime
+def databaseFetch(cursor,db,select,datetime):
+    sql="select noofbike from Bike_Detail where bikepointname= %s and datetime= %s"

```

Comment:
It would be difficult for the user to know the exact timestamp of update. Can you instead make a query that gives the latest row just before the timestamp mentioned by the user?

Reply:
Here  i am just asking the user to input the bike point name. And before query   i am taking the current local time into consideration and then i am querying the DB to fetch the releavant Data . User will only provide the bike point name and not the time.